### PR TITLE
Add `EmbeddingOntology`

### PR DESCRIPTION
**Pull request by capjamesg**
This PR adds an `EmbeddingOntology` for use with Autodistill. This ontology is designed for using a classification model with custom embeddings. The `EmbeddingOntology` should be used in the following way:

- Calculate a CLIP vector for an object representative of a given class
- Map the vector to a class name
- Pass the mapping between vectors and class names into the `EmbeddingOntology`

This ontology must be explicitly supported by a base model (i.e. CLIP) since it introduces new behavior (classification through vector comparison).

Here is an example code snippet that uses this ontology:

```python
from autodistill_clip import CLIP
from autodistill.detection import CaptionOntology
from autodistill_grounding_dino import GroundingDINO
from autodistill.core import EmbeddingOntology
from autodistill.core.custom_detection_model import CustomDetectionModel

import torch
import clip
from PIL import Image
import os

device = "cuda" if torch.cuda.is_available() else "cpu"
model, preprocess = clip.load("ViT-B/32", device=device)

images = os.listdir("samples")

images_to_classes = {
    "midnights": "IMG_9022.jpeg",
    "men amongst mountains": "323601467684.jpeg",
    "we are": "IMG_9056.jpeg",
    "oh wonder": "Images (5).jpeg",
    "brightside": "Images (4).jpeg",
    "tears for fears": "Images (3).jpeg"
}

embeddings_to_classes = {}

with torch.no_grad():
    for cls, image in images_to_classes.items():
        image = preprocess(Image.open(os.path.join("samples", image))).unsqueeze(0).to(device)
        embeddings_to_classes[tuple(model.encode_image(image).cpu().numpy()[0])] = cls


SAMCLIP = CustomDetectionModel(
    detection_model=GroundingDINO(
        CaptionOntology({"album cover": "album cover"})
    ),
    classification_model=CLIP(
        EmbeddingOntology(embeddings_to_classes.items())
    )
)

SAMCLIP.label(input_folder="./snap-images", output_folder="./dataset", extension=".jpeg")
```

### CODE REVIEW

**Comment by capjamesg** [comment]
We discussed this further and I took away the following action items:

- [ ] Base models that support embedding should inherit from an `EmbeddingModel` parent class, as well as the `BaseModel` class.
- [ ] Embedding models should support a `.embed_image` function, which embeds an image.
- [ ] The EmbeddingOntology should be {class_name:image}, where `image` can be a file name or a NumPy array. Alternatively, a list of file names or NumPy arrays can be provided. `.embed_image` will be called on each image; if a list is provided, infer will run on each image and an average will be taken.

**Comment by yeldarby** [comment]
`embed_image` should also probably have a `normalize` parameter that returns normalized vectors.

> The EmbeddingOntology should be {class_name:image}, where image can be a file name or a NumPy array. Alternatively, a list of file names or NumPy arrays can be provided. .embed_image will be called on each image; if a list is provided, infer will run on each image and an average will be taken.

Let's just simplify this by requiring it to be an `images` array (in the base case where you only want one image you can just pass a one-image array).

We might want to plan for a `cluster` param on `EmbeddingOntology` that'd give multiple pooled embeddings for each class (eg if you're passing in a whole dataset of 1000 images you could set `clusters=10` and for each class we'd create 10 clusters to use for the similarity calculation to cover a broader embedding space for the class in case there's lots of variability). For now we could just default clusters=1 and throw an error that it's not implemented if they try to pass another value.

**Comment by capjamesg** [comment]
@yeldarby See the updated `autodistill/core/embedding_ontology.py` file for my latest iteration. My idea is you would call `process()` in the base model `__init__` constructor to process the ontology, then the `infer()` method in each base model would reference the embedding ontology.

File: autodistill/core/embedding_model.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import numpy as np
+
+from .ontology import Ontology
+
+
+@dataclass
+class EmbeddingModel(ABC):
+    ontology: Ontology
+
+    def set_ontology(self, ontology: Ontology):
```
Comment (yeldarby) [inline autodistill/core/embedding_model.py:14]: I think we don't need this

**Review by yeldarby — COMMENTED** [review]
We talked about this & are going to rework it a bit more to split out to various EmbeddingOntology types

* `EmbeddingOntologyRaw` - you're responsible for computing & passing in the embeddings themselves
* `EmbeddingOntologyImage` - takes an `EmbeddingModel` and a mapping of arrays of images (files or image objects; using the new `load_image` helper to determine & load them in the right way) to output prompts & will use that model to embed the files & average their vectors. This is the one that will in the future let you cluster the embeddings first .
* `EmbeddingOntologyText` - similar to a CaptionOntology where you pass in strings but it converts them to embeddings to use.

Pulling in the cosine similarity & image embedding code into here so it's not duplicated everywhere that needs to use this. This takes an image and returns back a class.

`CLIP` should inherit form both `ClassificationBaseModel` and `EmbeddingModel` and then `EmbeddingOntology` is constructed with an `EmbeddingModel` which lets us get rid of `process` and do it at initialization.
